web/tests/compose_validate.test.js

```diff
@@ -854,3 +863,75 @@ test_ui("test warn_if_topic_resolved", ({override, mock_template}) => {
     compose_validate.warn_if_topic_resolved(false);
     assert.ok(!error_shown);
 });
+
+test_ui("test_warn_if_guest_in_dm_recipient", ({override_rewire, mock_template}) => {
+    let is_active = false;
+
+    mock_template("compose_banner/guest_in_dm_recipient_warning.hbs", false, (data) => {
+        assert.equal(data.classname, compose_banner.CLASSNAMES.guest_in_dm_recipient_warning);
+        assert.equal(
+            data.banner_text,
+            $t({defaultMessage: "Guest is a guest in this organization."}),
+        );
+        is_active = true;
+        return "<banner-stub>";
+    });
+
+    compose_state.set_message_type("private");
+    initialize_pm_pill(mock_template);
+    compose_state.private_message_recipient("guest@example.com");
+    const classname = compose_banner.CLASSNAMES.guest_in_dm_recipient_warning;
+    let $banner = $(`#compose_banners .${CSS.escape(classname)}`);
+
+    // if setting is disabled, remove warning if exists
+    realm.realm_enable_guest_user_dm_warning = false;
+    compose_validate.warn_if_guest_in_dm_recipient();
+    assert.ok(!is_active);
+
+    // show warning for guest emails, banner should be shown
+    realm.realm_enable_guest_user_dm_warning = true;
+    $banner.length = 0;
+    compose_validate.warn_if_guest_in_dm_recipient();
+    assert.ok(is_active);
+    assert.equal(
+        compose_state.get_recipient_guest_emails_for_dm_warning().join(","),
+        "guest@example.com",
+    );
+
+    // don't show warning for same guests if user closed the banner.
+    is_active = false;
+    compose_validate.warn_if_guest_in_dm_recipient();
+    assert.ok(!is_active);
+
+    // on modifying the guest recipient, update banner if already shown.
+    is_active = true;
+    const new_guest = {
+        email: "new_guest@example.com",
+        user_id: 34,
+        full_name: "New Guest",
+        is_guest: true,
+    };
+    people.add_active_user(new_guest);
+
+    initialize_pm_pill(mock_template);
+    override_rewire(ui_util, "place_caret_at_end", noop);
+    compose_state.private_message_recipient("guest@example.com, new_guest@example.com");
+    $banner = $(`#compose_banners .${CSS.escape(classname)}`);
+    $banner.length = 1;
+    let is_updated = false;
+    $banner.set_find_results(".banner_content", {
+        text(content) {
+            assert.equal(
+                content,
+                $t({defaultMessage: "Guest and New Guest are guests in this organization."}),
+            );
+            is_updated = true;
+        },
+    });
```

Comment:
Why do have two approaches for testing similar type of stuff here? For the first case, we use `mock_template` and for the second one `set_find_results`.

Reply:
1st case is rendering the banner. The 2nd case is for updating the already existing banner text.

![image](https://github.com/user-attachments/assets/892e6165-fb2c-4e79-8ef8-5fb0486a7fef)